core/vcpu.c

```diff
@@ -2012,606 +1946,132 @@ static bool is_mmio_address(struct vcpu_t *vcpu, paddr_t gpa)
-                       " gva=0x%llx (CS:IP=0x%llx:0x%llx), mmio_gpa=0x%llx\n",
-                       __func__, vcpu->vcpu_id, va, cs_base, rip, dc->gpa);
+            " gva=0x%llx (CS:IP=0x%llx:0x%llx)\n",
+            __func__, vcpu->vcpu_id, va, cs_base, rip);
         dump_vmcs(vcpu);
         return -1;
     }
 #else  // !CONFIG_HAX_EPT2
     if (!vcpu_read_guest_virtual(vcpu, va, &instr, INSTR_MAX_LEN, INSTR_MAX_LEN,
-                                 0)) {
+        0)) {
         hax_panic_vcpu(vcpu, "Error reading instruction at 0x%llx for decoding"
-                       " (CS:IP=0x%llx:0x%llx)\n", va, cs_base, rip);
+            " (CS:IP=0x%llx:0x%llx)\n", va, cs_base, rip);
         dump_vmcs(vcpu);
         return -1;
     }
 #endif  // CONFIG_HAX_EPT2
 
-    // See http://wiki.osdev.org/X86-64_Instruction_Encoding
-    default_16bit = !is_64bit_mode && ((vcpu->state->_cr0 & CR0_PE) == 0 ||
-                    vcpu->state->_cs.operand_size == 0);
-
-    // Parse legacy prefixes
-    dc->has_rep = false;
-    for (num = 0; num < INSTR_MAX_LEGACY_PF; num++) {
-        switch (instr[num]) {
-            case 0xf0: {
-                // LOCK prefix
-                // Ignored (is it possible to emulate atomic operations?)
-                break;
-            }
-            case 0xf3: {
-                // REP prefix
-                dc->has_rep = true;
-                break;
-            }
-            case 0x66: {
-                // Operand-size override prefix
-                override_operand_size = 1;
-                break;
-            }
-            case 0x67: {
-                // Address-size override prefix
-                override_address_size = 1;
-                break;
-            }
-            case 0x2e: {
-                // CS segment override prefix
-                override_segment = PF_SEG_OVERRIDE_CS;
-                break;
-            }
-            case 0x36: {
-                // SS segment override prefix
-                override_segment = PF_SEG_OVERRIDE_SS;
-                break;
-            }
-            case 0x3e: {
-                // DS segment override prefix
-                override_segment = PF_SEG_OVERRIDE_DS;
-                break;
-            }
-            case 0x26: {
-                // ES segment override prefix
-                override_segment = PF_SEG_OVERRIDE_ES;
-                break;
-            }
-            case 0x64: {
-                // FS segment override prefix
-                override_segment = PF_SEG_OVERRIDE_FS;
-                break;
-            }
-            case 0x65: {
-                // GS segment override prefix
-                override_segment = PF_SEG_OVERRIDE_GS;
-                break;
-            }
-            default: {
-                goto done_legacy_pf;
-            }
-        }
+    rc = em_decode_insn(em_ctx, instr);
+    if (rc != EM_CONTINUE) {
+        return HAX_EXIT;
```

Comment:
This function is part of the `HAX_VCPU_IOCTL_RUN` handler where `HAX_EXIT` actually indicates success. See my other comment above on error handling.